File: tests/cuda/python/test_device_info.py
"""Device information CUDA binding tests."""

import sys
import os
import pytest

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
except ImportError:
    pytest.skip("Insight not available", allow_module_level=True)

try:
    ins.init()
    if ins.device_count() == 0:
        pytest.skip("No GPU available", allow_module_level=True)
except Exception:
    pytest.skip("GPU backend not available", allow_module_level=True)


class TestDeviceInfoCUDA:
    def test_device_name_gpu(self):
        name = ins.device_name("gpu", 0)
        assert isinstance(name, str)
        assert len(name) > 0

    def test_active_gpu_backend_name(self):
        name = ins.active_gpu_backend_name()
        assert isinstance(name, str)
        assert len(name) > 0

    def test_device_name_rejects_unknown_kind(self):
        with pytest.raises(ValueError):
            ins.device_name(ins.active_gpu_backend_name(), 0)

    def test_gpu_version_positive(self):
        ver = ins.gpu_version()
        assert ver > 0

    def test_driver_version_positive(self):
        ver = ins.driver_version()
        assert ver > 0

    def test_device_count_positive(self):
        count = ins.device_count()
        assert count > 0

    def test_compute_capability_positive(self):
        cc = ins.compute_capability(0)
        assert cc > 0

    def test_device_memory(self):
        mem = ins.device_memory(0)
        assert isinstance(mem, tuple)
        assert len(mem) == 2
        assert mem[0] > 0  # total > 0
        assert mem[1] > 0  # free > 0
        assert mem[0] >= mem[1]  # total >= free

    def test_compute_capability_stays_positive(self):
        cc = ins.compute_capability(0)
        assert cc > 0

    def test_gpu_version_stays_positive(self):
        ver = ins.gpu_version()
        assert ver > 0

    def test_driver_version_stays_positive(self):
        ver = ins.driver_version()
        assert ver > 0

    def test_device_memory_total_reasonable(self):
        total, free = ins.device_memory(0)
        # At least 1 GB
        assert total >= 1024 * 1024 * 1024

    def test_device_name_multiple_calls(self):
        # Should be stable across calls
        name1 = ins.device_name("gpu", 0)
        name2 = ins.device_name("gpu", 0)
        assert name1 == name2

    def test_compute_capability_multiple_calls(self):
        cc1 = ins.compute_capability(0)
        cc2 = ins.compute_capability(0)
        assert cc1 == cc2

    def test_is_initialized(self):
        assert ins.is_initialized()

    def test_device_name_gpu_string(self):
        name = ins.device_name("gpu", 0)
        # Should contain GPU-related info
        assert len(name) > 0


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/cuda/lua/test_device_info.lua
-- Device information CUDA binding tests
-- Run with:
--   LUA_PATH="bindings/lua/?/init.lua;;" LUA_CPATH="build/bindings/lua/?.so;;" \
--   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda \
--   ~/.luarocks/bin/busted tests/cuda/lua/test_device_info.lua

local ok_ins, ins = pcall(require, "_insight")
if not ok_ins then
  print("SKIP: _insight module not available")
  os.exit(0)
end

local ok_gpu = pcall(function()
  ins.init()
end)
if not ok_gpu then
  print("SKIP: GPU backend not available")
  os.exit(0)
end

describe("DeviceInfo CUDA Tests", function()
  it("device_name gpu", function()
    local name = ins.device_name("gpu", 0)
    assert.is_not_nil(name)
    assert.is_string(name)
    assert.is_true(#name > 0)
  end)

  it("active_gpu_backend_name", function()
    local name = ins.active_gpu_backend_name()
    assert.is_string(name)
    assert.is_true(#name > 0)
  end)

  it("device_name rejects backend name", function()
    assert.has_error(function()
      ins.device_name(ins.active_gpu_backend_name(), 0)
    end)
  end)

  it("gpu_version positive", function()
    local ver = ins.gpu_version()
    assert.is_number(ver)
    assert.is_true(ver > 0)
  end)

  it("device_count positive", function()
    local count = ins.device_count()
    assert.is_number(count)
    assert.is_true(count > 0)
  end)

  it("compute_capability positive", function()
    local cc = ins.compute_capability(0)
    assert.is_number(cc)
    assert.is_true(cc > 0)
  end)

  it("compute_capability stays positive", function()
    local cc = ins.compute_capability(0)
    assert.is_true(cc > 0)
  end)

  it("gpu_version stays positive", function()
    local ver = ins.gpu_version()
    assert.is_true(ver > 0)
  end)

  it("device_name stable", function()
    local n1 = ins.device_name("gpu", 0)
    local n2 = ins.device_name("gpu", 0)
    assert.are.equal(n1, n2)
  end)

  it("compute_capability stable", function()
    local cc1 = ins.compute_capability(0)
    local cc2 = ins.compute_capability(0)
    assert.are.equal(cc1, cc2)
  end)

  it("device_name multiple calls", function()
    for i = 1, 5 do
      local name = ins.device_name("gpu", 0)
      assert.is_not_nil(name)
    end
  end)
end)
